### DISCUSSION TITLE
PROTOCOLO MODBUS TCP/IP ESP32-WROOM

### DISCUSSION THREAD

**Discussion by iiTzKyubiEVO** [community/community#131467] [Programming Help]
### Select Topic Area

Question

### Body

#include <WiFi.h>
#include <LiquidCrystal.h>
#include <ModbusRTU.h>

// Configuración de red
const char* ssid = "OXIMETRO PROTOCOLO TCP-IP"; // Reemplaza con el SSID de tu red
const char* password = "sdfg1234"; // Reemplaza con la contraseña de tu red

WiFiServer server(80); // Puerto del servidor TCP

// Conexiones de la pantalla LCD
const int rs = 14, en = 15, d4 = 16, d5 = 17, d6 = 18, d7 = 19;
LiquidCrystal lcd(rs, en, d4, d5, d6, d7);

// Pines del LED y botón
const int ledPin = 5;
const int buttonPin = 4;

// Configuración de Modbus RTU
ModbusRTU mb;
uint16_t modbusData[10]; // Datos Modbus

uint16_t handleModbus(TRegister* reg, uint16_t val) {
  // Ejemplo de manejo de datos Modbus
  modbusData[0] = digitalRead(buttonPin); // Estado del botón
  modbusData[1] = digitalRead(ledPin);    // Estado del LED
  return val;
}

void setup() {
  Serial.begin(115200);
  Serial2.begin(9600, SERIAL_8N1, 16, 17); // Configura Serial2 para Modbus

  // Inicializar la pantalla LCD
  lcd.begin(16, 2);
  lcd.setCursor(0, 0);
  lcd.print("Conectando...");

  // Configuración del pin del LED como salida
  pinMode(ledPin, OUTPUT);
  digitalWrite(ledPin, LOW); // Inicialmente apagado

  // Configuración del pin del botón como entrada con pull-up interno
  pinMode(buttonPin, INPUT_PULLUP);

  // Conexión a la red WiFi
  WiFi.begin(ssid, password);
  while (WiFi.status() != WL_CONNECTED) {
    delay(1000);
    Serial.println("Conectando a WiFi...");
    lcd.setCursor(0, 1);
    lcd.print("Conectando...");
  }

  Serial.println("Conectado a la red WiFi");
  Serial.print("Dirección IP: ");
  Serial.println(WiFi.localIP());

  // Inicia el servidor
  server.begin();
  Serial.println("Servidor TCP iniciado");

  lcd.clear();
  lcd.setCursor(0, 0);
  lcd.print("IP:");
  lcd.setCursor(3, 0);
  lcd.print(WiFi.localIP());

  // Configuración de Modbus
  mb.begin(&Serial2);
  mb.master();
  mb.addIreg(0, 0, 10); // Añade registros Modbus

  // Añadir manejadores Modbus
  mb.addHreg(100, 0);
  mb.onSetHreg(100, [](TRegister* reg, uint16_t val) -> uint16_t {
    if (reg->address.address == 100) { // Compara el valor del registro
      digitalWrite(ledPin, val); // Controla el LED mediante Modbus
    }
    return val;
  });

  mb.onGetIreg(0, handleModbus);
}

void loop() {
  // Ejecuta Modbus
  mb.task();

  // Comprueba si hay nuevos clientes
  WiFiClient client = server.available();
  if (client) {
    String request = client.readStringUntil('\r');
    client.flush();

    // Control del LED basado en la petición del cliente
    if (request.indexOf("/LED=ON") != -1) {
      digitalWrite(ledPin, HIGH); // Enciende el LED
    } else if (request.indexOf("/LED=OFF") != -1) {
      digitalWrite(ledPin, LOW); // Apaga el LED
    }

    // Lee el estado del botón
    int buttonState = digitalRead(buttonPin);
    String buttonStatus = (buttonState == LOW) ? "ACTIVADO" : "DESACTIVADO";

    // Analiza la solicitud para ver si contiene datos de texto
    if (request.indexOf("GET /text?message=") != -1) {
      int startIndex = request.indexOf('=') + 1;
      int endIndex = request.indexOf(' ', startIndex);
      if (endIndex == -1) {
        endIndex = request.indexOf('\r', startIndex);
      }
      String message = request.substring(startIndex, endIndex);
      message.replace("%20", " "); // Reemplaza los espacios codificados
      Serial.println("Mensaje recibido: " + message);

      // Muestra el mensaje en la pantalla LCD
      lcd.clear();
      lcd.setCursor(0, 0);
      lcd.print(message);
    }

    // Envío de la página web al cliente
    if (request.indexOf("/status") != -1) {
      client.println("HTTP/1.1 200 OK");
      client.println("Content-Type: text/plain");
      client.println();
      client.print(buttonStatus);
    } else {
      client.println("HTTP/1.1 200 OK");
      client.println("Content-Type: text/html");
      client.println();
      client.println("<!DOCTYPE HTML>");
      client.println("<html>");
      client.println("<head>");
      client.println("<title>Control de LCD y LED</title>");
      client.println("<meta charset=\"UTF-8\">");
      client.println("<meta name=\"viewport\" content=\"width=device-width, initial-scale=1\">");
      client.println("<script>");
      client.println("function updateStatus() {");
      client.println("  var xhr = new XMLHttpRequest();");
      client.println("  xhr.onreadystatechange = function() {");
      client.println("    if (this.readyState == 4 && this.status == 200) {");
      client.println("      document.getElementById('buttonStatus').innerHTML = this.responseText;");
      client.println("    }");
      client.println("  };");
      client.println("  xhr.open('GET', '/status', true);");
      client.println("  xhr.send();");
      client.println("}");
      client.println("setInterval(updateStatus, 1000);"); // Actualiza cada segundo
      client.println("</script>");
      client.println("</head>");
      client.println("<body>");
      client.println("<h1>Protocolo ModBus por Jimmy Quitral y Arturo Briceno!</h1>");
      client.println("<h2>Control del LED</h2>");
      client.println("<p><a href=\"/LED=ON\">Encender LED</a></p>");
      client.println("<p><a href=\"/LED=OFF\">Apagar LED</a></p>");
      client.println("<p>ESTADO DEL BOTON: <span id=\"buttonStatus\">Desconocido</span></p>");
      client.println("<h2>Enviar texto a la LCD</h2>");
      client.println("<form action=\"/text\" method=\"get\">");
      client.println("Texto: <input type=\"text\" name=\"message\"><br>");
      client.println("<input type=\"submit\" value=\"Enviar\">");
      client.println("</form>");
      client.println("</body>");
      client.println("</html>");
    }

    // Asegura que el cliente se desconecte correctamente
    client.stop();
  }
}

**Comment by MostlyKIGuess** [comment]
what is the question?

**Comment by ahmad13614** [comment]
> ### Select Topic Area
> 
> Question
> 
> ### Body
> 
> #include <WiFi.h>
> #include <LiquidCrystal.h>
> #include <ModbusRTU.h>
> 
> // Configuración de red
> const char* ssid = "OXIMETRO PROTOCOLO TCP-IP"; // Reemplaza con el SSID de tu red
> const char* password = "sdfg1234"; // Reemplaza con la contraseña de tu red
> 
> WiFiServer server(80); // Puerto del servidor TCP
> 
> // Conexiones de la pantalla LCD
> const int rs = 14, en = 15, d4 = 16, d5 = 17, d6 = 18, d7 = 19;
> LiquidCrystal lcd(rs, en, d4, d5, d6, d7);
> 
> // Pines del LED y botón
> const int ledPin = 5;
> const int buttonPin = 4;
> 
> // Configuración de Modbus RTU
> ModbusRTU mb;
> uint16_t modbusData[10]; // Datos Modbus
> 
> uint16_t handleModbus(TRegister* reg, uint16_t val) {
>   // Ejemplo de manejo de datos Modbus
>   modbusData[0] = digitalRead(buttonPin); // Estado del botón
>   modbusData[1] = digitalRead(ledPin);    // Estado del LED
>   return val;
> }
> 
> void setup() {
>   Serial.begin(115200);
>   Serial2.begin(9600, SERIAL_8N1, 16, 17); // Configura Serial2 para Modbus
> 
>   // Inicializar la pantalla LCD
>   lcd.begin(16, 2);
>   lcd.setCursor(0, 0);
>   lcd.print("Conectando...");
> 
>   // Configuración del pin del LED como salida
>   pinMode(ledPin, OUTPUT);
>   digitalWrite(ledPin, LOW); // Inicialmente apagado
> 
>   // Configuración del pin del botón como entrada con pull-up interno
>   pinMode(buttonPin, INPUT_PULLUP);
> 
>   // Conexión a la red WiFi
>   WiFi.begin(ssid, password);
>   while (WiFi.status() != WL_CONNECTED) {
>     delay(1000);
>     Serial.println("Conectando a WiFi...");
>     lcd.setCursor(0, 1);
>     lcd.print("Conectando...");
>   }
> 
>   Serial.println("Conectado a la red WiFi");
>   Serial.print("Dirección IP: ");
>   Serial.println(WiFi.localIP());
> 
>   // Inicia el servidor
>   server.begin();
>   Serial.println("Servidor TCP iniciado");
> 
>   lcd.clear();
>   lcd.setCursor(0, 0);
>   lcd.print("IP:");
>   lcd.setCursor(3, 0);
>   lcd.print(WiFi.localIP());
> 
>   // Configuración de Modbus
>   mb.begin(&Serial2);
>   mb.master();
>   mb.addIreg(0, 0, 10); // Añade registros Modbus
> 
>   // Añadir manejadores Modbus
>   mb.addHreg(100, 0);
>   mb.onSetHreg(100, [](TRegister* reg, uint16_t val) -> uint16_t {
>     if (reg->address.address == 100) { // Compara el valor del registro
>       digitalWrite(ledPin, val); // Controla el LED mediante Modbus
>     }
>     return val;
>   });
> 
>   mb.onGetIreg(0, handleModbus);
> }
> 
> void loop() {
>   // Ejecuta Modbus
>   mb.task();
> 
>   // Comprueba si hay nuevos clientes
>   WiFiClient client = server.available();
>   if (client) {
>     String request = client.readStringUntil('\r');
>     client.flush();
> 
>     // Control del LED basado en la petición del cliente
>     if (request.indexOf("/LED=ON") != -1) {
>       digitalWrite(ledPin, HIGH); // Enciende el LED
>     } else if (request.indexOf("/LED=OFF") != -1) {
>       digitalWrite(ledPin, LOW); // Apaga el LED
>     }
> 
>     // Lee el estado del botón
>     int buttonState = digitalRead(buttonPin);
>     String buttonStatus = (buttonState == LOW) ? "ACTIVADO" : "DESACTIVADO";
> 
>     // Analiza la solicitud para ver si contiene datos de texto
>     if (request.indexOf("GET /text?message=") != -1) {
>       int startIndex = request.indexOf('=') + 1;
>       int endIndex = request.indexOf(' ', startIndex);
>       if (endIndex == -1) {
>         endIndex = request.indexOf('\r', startIndex);
>       }
>       String message = request.substring(startIndex, endIndex);
>       message.replace("%20", " "); // Reemplaza los espacios codificados
>       Serial.println("Mensaje recibido: " + message);
> 
>       // Muestra el mensaje en la pantalla LCD
>       lcd.clear();
>       lcd.setCursor(0, 0);
>       lcd.print(message);
>     }
> 
>     // Envío de la página web al cliente
>     if (request.indexOf("/status") != -1) {
>       client.println("HTTP/1.1 200 OK");
>       client.println("Content-Type: text/plain");
>       client.println();
>       client.print(buttonStatus);
>     } else {
>       client.println("HTTP/1.1 200 OK");
>       client.println("Content-Type: text/html");
>       client.println();
>       client.println("<!DOCTYPE HTML>");
>       client.println("<html>");
>       client.println("<head>");
>       client.println("<title>Control de LCD y LED</title>");
>       client.println("<meta charset=\"UTF-8\">");
>       client.println("<meta name=\"viewport\" content=\"width=device-width, initial-scale=1\">");
>       client.println("<script>");
>       client.println("function updateStatus() {");
>       client.println("  var xhr = new XMLHttpRequest();");
>       client.println("  xhr.onreadystatechange = function() {");
>       client.println("    if (this.readyState == 4 && this.status == 200) {");
>       client.println("      document.getElementById('buttonStatus').innerHTML = this.responseText;");
>       client.println("    }");
>       client.println("  };");
>       client.println("  xhr.open('GET', '/status', true);");
>       client.println("  xhr.send();");
>       client.println("}");
>       client.println("setInterval(updateStatus, 1000);"); // Actualiza cada segundo
>       client.println("</script>");
>       client.println("</head>");
>       client.println("<body>");
>       client.println("<h1>Protocolo ModBus por Jimmy Quitral y Arturo Briceno!</h1>");
>       client.println("<h2>Control del LED</h2>");
>       client.println("<p><a href=\"/LED=ON\">Encender LED</a></p>");
>       client.println("<p><a href=\"/LED=OFF\">Apagar LED</a></p>");
>       client.println("<p>ESTADO DEL BOTON: <span id=\"buttonStatus\">Desconocido</span></p>");
>       client.println("<h2>Enviar texto a la LCD</h2>");
>       client.println("<form action=\"/text\" method=\"get\">");
>       client.println("Texto: <input type=\"text\" name=\"message\"><br>");
>       client.println("<input type=\"submit\" value=\"Enviar\">");
>       client.println("</form>");
>       client.println("</body>");
>       client.println("</html>");
>     }
> 
>     // Asegura que el cliente se desconecte correctamente
>     client.stop();
>   }
> }
